patroni/postgresql.py

```diff
@@ -296,12 +317,40 @@ def write_recovery_conf(self, leader):
                 for name, value in self.config.get('recovery_conf', {}).items():
                     f.write("{} = '{}'\n".format(name, value))
 
+    def prepare_pg_rewind_connection(self, leader_url, pg_rewind):
+        r = parseurl(leader_url)
+        r.update(pg_rewind)
+        env = self.write_pgpass(r, append=True)
+        return (env, "user={user} host={host} port={port} dbname=postgres sslmode=prefer sslcompression=1".format(**r))
+
+    def pg_rewind(self, leader):
+        env, pc = self.prepare_pg_rewind_connection(leader.conn_url, self._pg_rewind)
+        logger.info("running pg_rewind from {}".format(pc))
+        pg_rewind = ['pg_rewind', '-D', self.data_dir, '--source-server', pc]
+        try:
+            ret = (subprocess.call(pg_rewind, env=env) == 0)
+        except:
+            ret = False
+        if ret:
+            self.write_recovery_conf(leader)
+        return ret
+
     def follow_the_leader(self, leader):
         if not self.check_recovery_conf(leader):
             self.write_recovery_conf(leader)
-            run_callback = self.role == 'master'
-            self.restart()
-            run_callback and self.call_nowait(ACTION_ON_ROLE_CHANGE)
+            change_role = self.role == 'master'
+
+            if leader and change_role and self._pg_rewind_present:
+                self.stop()
+                if self.pg_rewind(leader):
+                    ret = self.start()
+                else:
+                    ret = False
+                    self.remove_data_directory()
```

Comment:
Removing the old data directory might be dangerous, because the former master might have some data not propagated to the new one (in the case of the split brain). Consider moving the data, or have a configuration option that defines what to do.